File: data-structure/02_queue/priority_queue.py
class PriorityQueue:

    def __init__(self):
        self._heap = list()
        self._heap.append(None)

    def _swap(self, a, b):
        self._heap[a], self._heap[b] = self._heap[b], self._heap[a]

    def is_empty(self):
        return len(self._heap) <= 1

    def put(self, data):
        self._heap.append(data)
        child_index = len(self._heap) - 1
        parent_index = child_index // 2

        while parent_index:
            if self._heap[child_index][0] < self._heap[parent_index][0]:
                self._swap(child_index, parent_index)
            child_index = parent_index
            parent_index //= 2
        return True

    def get(self):
        if self.is_empty():
            return None

        _, data = self._heap[1]
        self._heap[1] = self._heap[-1]
        del self._heap[-1]

        length = len(self._heap) - 1

        parent_index = 1
        left_child_index = parent_index * 2
        right_child_index = parent_index * 2 + 1

        def swap_down(parent_index, child_index):
            self._swap(parent_index, child_index)
            parent_index = child_index
            left_child_index = parent_index * 2
            right_child_index = parent_index * 2 + 1
            return parent_index, left_child_index, right_child_index

        while length >= left_child_index:
            if length < right_child_index:
                if self._heap[left_child_index][0] < self._heap[parent_index][0]:
                    parent_index, left_child_index, right_child_index = \
                        swap_down(parent_index, left_child_index)
                else:
                    break
            elif self._heap[left_child_index][0] <= self._heap[right_child_index][0]:
                parent_index, left_child_index, right_child_index = \
                    swap_down(parent_index, left_child_index)
            else:
                parent_index, left_child_index, right_child_index = \
                    swap_down(parent_index, right_child_index)

        return data

    def peek(self):
        return self._heap[1][1]


if __name__ == "__main__":
    pq = PriorityQueue()
    pq.put((0, 'a'))
    pq.put((5, 'b'))
    pq.put((2, 'c'))
    pq.put((1, 'd'))
    pq.put((3, 'e'))
    pq.put((4, 'f'))
    print(pq._heap)
    print(pq.get())
    print(pq.get())
    print(pq.get())
    print(pq.get())
    print(pq.get())
    print(pq.get())
    print(pq.get())

File: data-structure/02_queue/PriorityQueue.js
class MinHeap {
  /**
   * @param {function(*, *): boolean} [comparator=(lhs, rhs) => lhs < rhs] sorting criteria
   */
  constructor(comparator = (lhs, rhs) => lhs < rhs) {
    this.heap = [0];
    this._comparator = comparator;
  }

  swap(a, b) {
    [this.heap[a], this.heap[b]] = [this.heap[b], this.heap[a]];
  }

  empty() {
    return this.heap.length === 1;
  }

  insert(data) {
    this.heap.push(data);
    const moveUp = (index = this.heap.length - 1) => {
      const parentIndex = Math.floor(index / 2);
      const parentData = this.heap[parentIndex];
      const currentData = this.heap[index];
      if (index === 1 || !this._comparator(currentData, parentData)) return;
      this.swap(index, parentIndex);
      moveUp(parentIndex);
    };
    moveUp();
  }

  extract() {
    if (this.heap.length <= 1) return;
    if (this.heap.length === 2) return this.heap.pop();
    const data = this.heap[1];
    this.heap[1] = this.heap.pop();

    const moveDown = (index = 1) => {
      const leftChildIndex = index * 2;
      const rightChildIndex = index * 2 + 1;
      const currentData = this.heap[index];
      const leftChildData = this.heap[leftChildIndex];
      const rightChildData = this.heap[rightChildIndex];
      if (!leftChildData) return;
      if (!rightChildData) {
        if (this._comparator(leftChildData, currentData)) {
          this.swap(leftChildIndex, index);
          moveDown(leftChildIndex);
        }
      } else {
        if (this._comparator(leftChildData, rightChildData)) {
          if (this._comparator(leftChildData, currentData)) {
            this.swap(leftChildIndex, index);
            moveDown(leftChildIndex);
          }
        } else {
          if (this._comparator(rightChildData, currentData)) {
            this.swap(rightChildIndex, index);
            moveDown(rightChildIndex);
          }
        }
      }
    };
    moveDown();

    return data;
  }
}

class PriorityQueue extends MinHeap {
  empty() {
    return this.heap.length === 1;
  }

  enqueue(data) {
    this.insert(data);
  }

  dequeue() {
    return this.extract();
  }
}

const priorityQueue = new PriorityQueue((a, b) => a[1] < b[1]);

priorityQueue.enqueue([10, 1]);
priorityQueue.enqueue([15, 2]);
priorityQueue.enqueue([7, 10]);
priorityQueue.enqueue([1, 77]);
priorityQueue.enqueue([5, 7]);
priorityQueue.enqueue([100, 30]);

console.log(priorityQueue.dequeue());
console.log(priorityQueue.dequeue());
console.log(priorityQueue.dequeue());

priorityQueue.enqueue([4, 40]);
priorityQueue.enqueue([3, 3]);
priorityQueue.enqueue([2, 7]);

console.log(priorityQueue.dequeue());
console.log(priorityQueue.dequeue());
console.log(priorityQueue.dequeue());

priorityQueue.enqueue([2, 100]);
priorityQueue.enqueue([2, 40]);
priorityQueue.enqueue([2, 1]);

console.log(priorityQueue.dequeue());
console.log(priorityQueue.dequeue());
console.log(priorityQueue.dequeue());
console.log(priorityQueue.dequeue());
console.log(priorityQueue.dequeue());
console.log(priorityQueue.dequeue());
console.log(priorityQueue.dequeue());
console.log(priorityQueue.dequeue());
